### PR TITLE
Allow first-match module name matching

### PR DESCRIPTION
**Pull request by salkinium**
This adds the ability to query whether modules with a certain prefix exist: `env.has_module("repo:name*")`.
The implementation of `ModuleNameResolver` just returns the first match if any.

- [x] initial implementation
- [ ] generic implementation
- [ ] unit tests

cc @dergraaf

### CODE REVIEW

**Comment by salkinium** [comment]
[Example](https://github.com/modm-io/modm/commit/65c64545f2608387e39f0b4adcaccd67bebf4ee5#diff-a885160f894446aa364d392ee7905bedR95):

```py
def build(env):
    if not (env.has_module(":board.nucleo-*") or env.has_module(":board.al-avr*")):
        env.log.warn("GPIO test has been hardcoded to a Nucleo-64 board. "
                     "When porting make sure this test does not damage your board!")
        return
```

**Comment by dergraaf** [comment]
Sounds reasonable. Would it make sense to implement a more general wildcard scheme?

File: lbuild/module.py [hunk 1]
```diff
@@ -236,6 +236,10 @@ def __getitem__(self, key: str):
         try:
             name = self.module.fill_partial_name(partial_name)
             key = ":".join(name)
+            find_first = (key[-1] == '*')
```
Comment (dergraaf) [inline lbuild/module.py:239]: This is not used, but should probably used to enable the following lines

File: lbuild/module.py [hunk 2]
```diff
@@ -236,6 +236,10 @@ def __getitem__(self, key: str):
...
+            if key not in self.modules:
+                keys = [k for k in self.modules if k.startswith(key[:-1])]
+                if len(keys): key = keys[0];
```
Comment (dergraaf) [inline lbuild/module.py:242]: Please don't use this style. There is (almost) no need for semicolons in python :smile:

**Review by dergraaf — CHANGES_REQUESTED** [review]

**Comment by salkinium** [comment]
Yeah, this was more of a quick hack, I wanted to start a discussion.

>  Would it make sense to implement a more general wildcard scheme?

Yes, absolutely! I was thinking something with a glob-like syntax, since we already have `:**` "any module" match. So something like `**:board.*` would collect *any* modules with a `board.` prefix.

**Pull request closed without merging**